src/manifold/src/quickhull.h

```diff
@@ -129,51 +130,28 @@ struct Ray {
       : S(S), V(V), VInvLengthSquared(1 / (glm::dot(V, V))) {}
 };
 
-// VertexDataSource
-
-class VertexDataSource {
-  const glm::dvec3* ptr;
-  size_t count;
-
- public:
-  VertexDataSource(const glm::dvec3* ptr, size_t count)
-      : ptr(ptr), count(count) {}
-
-  VertexDataSource(const std::vector<glm::dvec3>& vec)
-      : ptr(&vec[0]), count(vec.size()) {}
-
-  VertexDataSource() : ptr(nullptr), count(0) {}
-
-  VertexDataSource& operator=(const VertexDataSource& other) = default;
-
-  size_t size() const { return count; }
-
-  const glm::dvec3& operator[](size_t index) const { return ptr[index]; }
-
-  const glm::dvec3* begin() const { return ptr; }
-
-  const glm::dvec3* end() const { return ptr + count; }
-};
-
 // Mesh.hpp
 
 class MeshBuilder {
  public:
   struct HalfEdge {
-    size_t endVertex;
-    size_t opp;
-    size_t face;
-    size_t next;
-
-    void disable() { endVertex = std::numeric_limits<size_t>::max(); }
+    // size_t endVertex;
+    // size_t opp;
+    // size_t face;
+    manifold::Halfedge halfEdgeManifold;
```

Comment:
If we can't actually use our `halfedge_` vector, you can go ahead and put this back the way it was so the members are easier to access.

Reply:
So, I had an idea regarding the same, I understand that we can't just use the `halfedge_` vector during each iteration, but at the end, I can reorder the halfedges and return our `halfedge_` vector. That would be a O(n) operation, so wouldn't add much overhead as well, and we will have the halfedges in our format if we want to do some further operation. What do you think about this? (Also another option is we can modify our halfedge implementation and add next in it, which I will only use in the quickhull file for now, it would still add a unused variable at every other place halfedge is being used, which does not seem ideal)

If you think we should implement any of these ideas, I'll go ahead with it otherwise I'll revert the changes

Reply:
If you want to do reordering, you should probably just have a vector of `manifold::Halfedge` as well as a vector of `int` as the `next` index.

Reply:
> If you want to do reordering, you should probably just have a vector of `manifold::Halfedge` as well as a vector of `int` as the `next` index.

Yes, I will remove the class and move the functions outside as well. I just wanted to get the idea approved first. Let's see what Emmett says, till then I'll see if I can start profiling

Reply:
Oh, yes, thanks for bringing this up - I just reread how your function in `manifold.cpp` works. Yes, let's make quickhull into a member function of `Impl` and directly construct the `halfedge_` vector. The public Manifold function can just wrap that. You can use [this](https://github.com/elalish/manifold/blob/master/src/manifold/src/impl.cpp#L485-L532) as an example. The halfedge reordering part can remain as an internal detail of the quickhull function.

Reply:
So, just to verify if I understood it correct, I want to create a function like this in the file

> Manifold Manifold::Impl::Quickhull(const std::vector<glm::dvec3>& points) {

inside this function I wall call the quickhull function, which would return a `halfedge_` vector and also return the convexhull right, and if that's the case how do we plan to use the `halfedge_` vector here? [Also, I will be reordering the edges inside the algorithm after it computes the hull]

Reply:
The Impl is the internal mesh data structure of Manifold. You'll make a new Impl constructor that creates `vertPos_` and `halfedge_` because that is what represents a Manifold mesh - the convex hull in this case. If you look at how the Manifold constructors work you'll see they just mostly wrap `Impl` methods. Then you won't need to use `Mesh` at all, which is good because we're deprecating it.

Reply:
I'm sorry but I'm still slightly confused. I can directly get vertPos and halfedge from the quickhull function, so why would I need to create a new Impl constructor for that? If you don't mind could write a pseudo code or function heading to explain it a bit.

Reply:
My understanding of this code right now is that `QuickHull` operates internally on a `MeshBuilder` object called `mesh` - this is where the algorithmic work actually happens. Once complete, it goes through a series of transforms: `MeshBuilder -> ConvexHull -> Mesh -> Manifold`, where `Manifold` is really just a wrapper of `Impl` which is our actual internal halfedge data structure. I propose simplifying this into `MeshBuilder -> Impl` - delete the `ConvexHull` class entirely and don't use our `Mesh` class at all. Simply translate the `MeshBuilder` halfedge data structure into our `Impl` halfedge data structure.

This can either happen in your existing `Manifold Manifold::Hull(const std::vector<vec3>& pts)` function, or in a new `void Manifold::Impl::Hull(const std::vector<vec3>& pts)` function - the latter just lets you avoid writing a lot of `impl_->...` but we have plenty of examples both ways, so I don't have a strong opinion.

Reply:
Thank you for the clarification, I understand it now. The last thing I wanted to ask is once, I translate the `MeshBuilder` halfedge data structure into `Impl` halfedge, then we plan to call functions like this https://github.com/elalish/manifold/blob/952546b761b343a21a951ff03b22a8d0a8a570f8/src/manifold/src/impl.cpp#L528C1-L531C17, so we can then directly use this 'Impl` as a wrapper function right?

Reply:
exactly - you can probably skip CreateFaces though - it's a bit slow and not so necessary on a hull.

Reply:
Alright, thank you for clearing it up, I was quite confused about what I had to do, but I think I've got a very good idea of it now.